src/base/RegexPattern.h

```diff
@@ -26,33 +24,29 @@ class RegexPattern
 
 public:
     RegexPattern() = delete;
-    RegexPattern(const SBuf &aPattern, int aFlags);
-    ~RegexPattern();
-
     RegexPattern(RegexPattern &&) = delete; // no copying of any kind
+    RegexPattern(const SBuf &aPattern, const std::regex::flag_type aFlags);
+    ~RegexPattern() = default;
 
     /// whether the regex differentiates letter case
-    bool caseSensitive() const { return !(flags & REG_ICASE); }
+    bool caseSensitive() const { return !(regex.flags() & std::regex::icase); }
 
     /// whether this is an "any single character" regex (".")
     bool isDot() const { return pattern.length() == 1 && pattern[0] == '.'; }
 
-    bool match(const char *str) const {return regexec(&regex,str,0,nullptr,0)==0;}
+    bool match(const char *) const;
 
     /// Attempts to reproduce this regex (context-sensitive) configuration.
     /// If the previous regex is nil, may not report default flags.
     /// Otherwise, may not report same-as-previous flags (and prepends a space).
     void print(std::ostream &os, const RegexPattern *previous = nullptr) const;
 
 private:
-    /// a regular expression in the text form, suitable for regcomp(3)
+    /// a regular expression in the text form
     SBuf pattern;
 
-    /// bitmask of REG_* flags for regcomp(3)
-    const int flags;
-
-    /// a "compiled pattern buffer" filled by regcomp(3) for regexec(3)
-    regex_t regex;
+    /// a "compiled pattern buffer"
+    std::regex regex;
```

Comment:
Latest constructor is changing it after construction.

Reply:
> Latest constructor is changing it after construction.

Yes, but only because the current implementation did not follow my explicit [suggestion](https://github.com/squid-cache/squid/pull/1232#r1064218193) on how to avoid "changing it after construction" AFAICT. Please try my suggestion and see if that allows us to make this data member constant. I will quote it here of convenience:

> I believe this reporting can be easily done by creating the regex in a simple helper function that takes the same arguments as the RegexPattern constructor, catches STL exceptions and converts them into a TextException with the pattern detail

The catching and throwing code exists already. Please let me know if I should try making the corresponding adjustments instead.

Reply:
I saw the suggestion, I do not see a private member being const as more beneficial than avoiding the +1 data copy for function return. Especially since the object itself is apparently complex enough to require a PoolingAllocator.

Reply:
> I saw the suggestion, I do not see a private member being const as more beneficial than avoiding the +1 data copy for function return. 

C++ does not work that way. In relevant cases, the suggested code results in _fewer_ operations than the current PR code. See copy elision [documentation](https://en.cppreference.com/w/cpp/language/copy_elision) for technical details.

* Without compiler optimizations (`-O0`; an unimportant in this performance discussion but possibly easier to understand case), the current PR code creates two std::regex objects and then assigns one to another. The suggested code creates just one std::regex object and never uses the std::regex assignment, copy or move operators -- this copy elision is perfectly legal since C++11 and required since C++17.
* `-O2` compilations of the current PR code remove most of the above overhead, but still contain such extra operations as creation and copying of an additional std::locale object (used as an std::regex data member)! The assembly code of the simplest PR-like version of RegexPattern constructor is 128 lines. Its suggested replacement equivalent is only 71 lines long. Blindly counting assembly lines is silly, and I am not an assembly expert, but I did look through those lines, and believe that the primary difference in these two numbers is a good illustration of what better C++ code might yield in terms of the overhead reduction.

<!-- https://cpp.godbolt.org/z/W1d8xqTxf -->

We should not be debating the benefits of member initialization versus assignment in such a basic context.

----

> Especially since the object itself is apparently complex enough to require a PoolingAllocator.

This change request is about std::regex objects. Those objects do not use a PoolingAllocator (unfortunately). However, you are correct that std::regex should be treated as "complex enough" to stimulate writing copying-avoidance code, such as the code suggested in this change request.

Reply:
FTR: with GCC 12.4:
```
  error: moving a local object in a return statement prevents copy elision [-Werror=pessimizing-move]
```

Reply:
FWIW, I agree with the compiler -- we should not be moving any local objects here. Please let me know if you want me to address this change request (based on the earlier suggestion).